File: src/main/java/io/pivotal/pal/tracker/InMemoryTimeEntryRepository.java
package io.pivotal.pal.tracker;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class InMemoryTimeEntryRepository implements TimeEntryRepository {
  private long id = 1;
  private Map<Long, TimeEntry> db = new HashMap<>();

  @Override
  public TimeEntry create(TimeEntry timeEntry) {
    timeEntry.setId(id++);
    db.put(timeEntry.getId(), timeEntry);
    return timeEntry;
  }

  @Override
  public TimeEntry update(long id, TimeEntry timeEntry) {
    TimeEntry found = find(id);
    if ( found != null ) {
      found.copyFrom(timeEntry);
      return found;
    }
    return null;
  }

  @Override
  public TimeEntry delete(long id) {
    return db.remove(id);
  }

  @Override
  public TimeEntry find(long id) {
    return db.get(id);
  }

  @Override
  public List<TimeEntry> list() {
    return new ArrayList<>(db.values());
  }
}

File: src/main/kotlin/io/pivotal/pal/tracker/InMemoryTimeEntryRepository.kt
package io.pivotal.pal.tracker

import java.util.*

class InMemoryTimeEntryRepository : TimeEntryRepository {
    private var id:Long = 1
    private val db:MutableMap<Long, TimeEntry> = HashMap()

    override fun create(timeEntry: TimeEntry): TimeEntry {
        timeEntry.id = id++
        db.put(timeEntry.id, timeEntry)
        return timeEntry
    }

    override fun find(id: Long): TimeEntry? {
        return db.get(id)
    }

    override fun list(): List<TimeEntry> {
        return db.values.toList()
    }

    override fun update(id: Long, timeEntry: TimeEntry): TimeEntry? {
        val update = find(id)
        if (update != null) {
            update.projectId = timeEntry.projectId
            update.userId = timeEntry.userId
            update.date = timeEntry.date
            update.hours = timeEntry.hours
            return update
        }
        return null
    }

    override fun delete(id: Long): TimeEntry? {
        return db.remove(id)
    }
}